pkg/storage/tsdb/tombstones.go

```diff
@@ -0,0 +1,360 @@
+
+}
+
+func (m *TombstoneManager) RemoveCancelledStateIfExists(ctx context.Context, requestID string) error {
+	exists, err := m.TombstoneExists(ctx, requestID, StateCancelled)
+	if err != nil {
+		level.Error(m.logger).Log("msg", "unable to check if the request has previously been cancelled", "requestID", requestID, "err", err)
+		return err
+	}
+
+	if exists {
+		if err = m.DeleteTombstoneFile(ctx, requestID, StateCancelled); err != nil {
+			level.Error(m.logger).Log("msg", "unable to delete tombstone with previously cancelled state", "requestID", requestID, "err", err)
+			return err
+		}
+		level.Info(m.logger).Log("msg", "Removing tombstone file with previously cancelled state", "requestID", requestID, "err", err)
+
+	}
+	return nil
+}
+
+func GetStateAndRequestIDFromTombstonePath(tombstonePath string) (string, BlockDeleteRequestState, error) {
+	// The filename of the request should be <requestID>.<state>.json
+
+	// This should get the first extension which is .json
+	filenameExtesion := filepath.Ext(tombstonePath)
+	filenameWithoutJSON := tombstonePath[0 : len(tombstonePath)-len(filenameExtesion)]
+
+	stateExtension := filepath.Ext(filenameWithoutJSON)
+	requestID := filenameWithoutJSON[0 : len(filenameWithoutJSON)-len(stateExtension)]
+
+	// Ensure that the state exists as the filename extension
+	if len(stateExtension) == 0 {
+		return "", "", ErrInvalidDeletionRequestState
+	}
+
+	state := BlockDeleteRequestState(stateExtension[1:])
+	if !isValidDeleteRequestState(state) {
+		return "", "", errors.Wrapf(ErrInvalidDeletionRequestState, "Filename extension is invalid for tombstone: %s", tombstonePath)
+
+	}
+
+	return requestID, state, nil
+}
+
+func ParseMatchers(selectors []string) ([]*labels.Matcher, error) {
+	// Convert the string selectors to label matchers
+	var m []*labels.Matcher
+
+	for _, selector := range selectors {
+		parsed, err := parser.ParseMetricSelector(selector)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error parsing metric selector")
+		}
+		//keep the matchers in a 1D slice because the deletions are applied based
+		// on the "and" between all matchers.
+		m = append(m, parsed...)
+	}
+
+	return m, nil
+}
+
+func (t *Tombstone) GetFilename() string {
+	return t.RequestID + "." + string(t.State) + ".json"
+}
+
+func (t *Tombstone) IsOverlappingInterval(minT int64, maxT int64) bool {
+	return t.StartTime <= maxT && minT < t.EndTime
+}
+
+func (t *Tombstone) GetCreateTime() time.Time {
+	return time.Unix(t.RequestCreatedAt/1000, 0)
+}
+
+func getTombstoneFileName(requestID string, state BlockDeleteRequestState) string {
+	return requestID + "." + string(state) + ".json"
+}
+
+func isValidDeleteRequestState(state BlockDeleteRequestState) bool {
+	switch state {
+	case
+		StatePending,
+		StateProcessed,
+		StateCancelled:
+		return true
+	}
+	return false
+}
+
+func (s BlockDeleteRequestState) GetStateOrder() (int, error) {
```

Comment:
I don't think it's a wise idea to export this method. This looks a lot like implementation detail that should not be exposed. I don't think we want anyone outside the `tsdb` package to depend on the fact that the "order" of `StatePending` is 0, "StateProcessed" is 2, etc. right? Because what is important is the relation between the "order" not the absolute value right? 

If we really want to expose something, consider  functions like `IsGreaterThan(a, b BlockDeleteRequestState)`, but I don't think we should in this case.

Reply:
Not sure why it was exported in the first place. I don't think the logic is required anywhere else. I made it private now.